File: PLP/JogosONPLP/src/Prolog/CarregaArquivos.pl
listaAvaliacoes(Path, ListaAvaliacoes) :-
    carregarAvaliacoes(Path, ListaBruta),
    processarListaBrutaAvaliacoes(ListaBruta, ListaAvaliacoes).

processarListaBrutaAvaliacoes([], []).
processarListaBrutaAvaliacoes([Atual|XS], ListaAvaliacoes) :-
    split_string(Atual, "|", " | ", AtualDividido),
    nth0(0, AtualDividido, Nome),
    nth0(1, AtualDividido, Jogo),
    nth0(2, AtualDividido, NotaBruta),
    atom_number(NotaBruta, Nota),
    
    nth0(3, AtualDividido, Comentario),
    construirAvaliacao(Nome, Jogo, Nota, Comentario, AvaliacaoAtual),
    processarListaBrutaAvaliacoes(XS, ListaProx),
    append([AvaliacaoAtual], ListaProx, ListaAvaliacoes).

carregarAvaliacoes(Path, Avaliacoes) :-
    open(Path, read, Stream),
    read_file(Stream,ListaAvaliacoesString),
    Avaliacoes = ListaAvaliacoesString,
    close(Stream).

listaJogos(Path, ListaJogos) :-
    carregarJogos(Path, ListaBruta),
    processarListaBrutaJogos(ListaBruta, ListaJogos).

processarListaBrutaJogos([], []).
processarListaBrutaJogos([Atual|XS], ListaJogos) :-
    split_string(Atual, "|", " | ", AtualDividido),

    nth0(0, AtualDividido, Nome),
    
    nth0(1, AtualDividido, CategoriasBruto),
    split_string(CategoriasBruto, ",", ", ", Categorias),

    nth0(2, AtualDividido, ReqMinimosBruto),
    split_string(ReqMinimosBruto, ",", ", ", ReqMinimos),

    nth0(3, AtualDividido, Plataforma),
    
    nth0(4, AtualDividido, PrecoBruto),
    atom_number(PrecoBruto, Preco),

    nth0(5, AtualDividido, IsOnlineBruto),
    atom_number(IsOnlineBruto, IsOnline),

    nth0(6, AtualDividido, AnoLancamentoBruto),
    atom_number(AnoLancamentoBruto, AnoLancamento),

    contruirJogo(Nome, Categorias, ReqMinimos, Plataforma, Preco, IsOnline, AnoLancamento, JogoAtual),
    processarListaBrutaJogos(XS, ListaProx),
    append([JogoAtual], ListaProx, ListaJogos).

carregarJogos(Path, Jogos) :-
    open(Path, read, Stream),
    read_file(Stream,ListaJogosString),
    Jogos = ListaJogosString,
    close(Stream).

listaUsuarios(Path, ListaUsuarios) :-
    carregarUsuarios(Path, ListaBruta),
    processarListaBrutaUsuarios(ListaBruta, ListaUsuarios).

processarListaBrutaUsuarios([], []).
processarListaBrutaUsuarios([Atual|XS], ListaUsuarios) :-
    construirUsuario(Atual, UsuarioAtual),
    processarListaBrutaUsuarios(XS, ListaProx),
    append([UsuarioAtual], ListaProx, ListaUsuarios).


carregarUsuarios(Path, Usuarios) :-
    open(Path, read, Stream),
    read_file(Stream,ListaUsuarios),
    Usuarios = ListaUsuarios,
    close(Stream).

File: PLP/JogosONPLP/src/Haskell/CarregaArquivos.hs
module CarregaArquivos where
    import Util as Util
    import Jogo as Jogo
    import Usuario as Usuario
    import Avaliacao as Avaliacao
    import Data.List.Split (splitOn)
    import qualified System.IO.Strict as Strict

    -- / Le o arquivo de dados de usuarios, e retorna uma lista de string que cada uma armazena um usuario.
    lerArquivoUsuarios :: String -> IO([String])
    lerArquivoUsuarios path = do
        arquivo <- Strict.readFile path
        let listaUsuarios = lines arquivo
        return listaUsuarios

    carregarUsuarios :: [String] -> [Usuario]
    carregarUsuarios lista = [decodeUsuario x | x <- lista]

    decodeUsuario :: String -> Usuario
    decodeUsuario line = Usuario {
                            Usuario.nickname = line
    }

    -- / Le o arquivo de dados de avaliações, e retorna uma lista de string que cada uma armazena uma avaliação.
    lerArquivoAvaliacoes :: String -> IO([String])
    lerArquivoAvaliacoes path = do
        arquivo <- Strict.readFile path
        let listaAvaliacoes = lines arquivo
        return listaAvaliacoes
    
    -- / Retorna uma lista com as avaliações a partir de uma lista de strings contendo as avaliações.
    carregarAvaliacoes :: [String] -> [Avaliacao]
    carregarAvaliacoes lista = [decodeAvaliacao x | x <- lista]
    
    -- / A partir da string gerada da avaliação, retorna a Avaliação.
    decodeAvaliacao :: String -> Avaliacao
    decodeAvaliacao line = Avaliacao {
                                Avaliacao.usuario = paramAvaliacao!!0,
                                Avaliacao.jogo = paramAvaliacao!!1,
                                Avaliacao.nota = read (paramAvaliacao!!2) :: Double,
                                Avaliacao.comentario = paramAvaliacao!!3
                            }
                        where paramAvaliacao = splitOn "|" line

    -- / Le o arquivo de dados de jogos, e retorna uma lista de strings que armazena cada jogo.
    lerArquivoJogos :: String -> IO([String])
    lerArquivoJogos path = do
        arquivo <- Strict.readFile path
        let listaJogos = lines arquivo
        return listaJogos

    -- / Retorna uma lista com os jogos a partir de uma lista de strings contendo os jogos.
    carregarJogos :: [String] -> [Jogo]
    carregarJogos lista = [decodeJogo x | x <- lista]

    -- / A partir da string gerada do jogo, retorna o Jogo.
    decodeJogo :: String -> Jogo
    decodeJogo line = Jogo {
                        Jogo.nome = paramJogo!!0,
                        Jogo.categorias = Util.parseStringList (paramJogo!!1),
                        Jogo.reqMinimos = Util.parseStringList (paramJogo!!2),
                        Jogo.plataforma = paramJogo!!3,
                        Jogo.preco = read (paramJogo!!4) :: Double,
                        Jogo.online = if (paramJogo!!5 == "True") then True else False,
                        Jogo.anoLancamento = read (paramJogo!!6) :: Integer
                    }
                where paramJogo = splitOn "|" line